[Memento] Add PT2 to Memory Snapshot

sraikund16 opened the pull request:
Summary:
To add PT2 information to memory snapshot we piggyback off of the Kineto implementation using record_function similar to adding the user annotations. To do this we add the following:

1. Stack implementation that we instantiate to keep track of which compile context stack we are currently in (top element of the stack). The stack will be per device and thread-local since different threads of a process can be in different compile contexts at a given time. For this reason, we do not need to add mutexes to our stack impl since no two threads will touch a given stack
2. RecordFunction hooks to properly pipe the correct events to the compile context stack. These hooks are similar to the annotation ones in the fact that we just register them lazily and DO NOT unregister them. This is done out of convenience. In the future, we should save the handles and unregister them to minimize overhead after profiling is finished. As of now, we are registering this at the FUNCTION scope which is wide; however, we treat any function that does not start with "Torch-Compiled Region" as a no-op so we anticipate the difference in performance to be negligible during and after profiling. We also hide this feature behind a flag set to off on default so existing jobs will be unaffected
3. Piping for compile context to pickle output

Test Plan:
In D74039793, we add CompileContext to the visualizer and we see the following {F1977654658}

Differential Revision: D74028214




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

sraikund16 commented on torch/csrc/cuda/memory_snapshot.cpp:
```diff
@@ -117,6 +117,36 @@ void _initRecordAnnotations() {
   }();
 }
 
+void _initCompileContexts() {
```
wondering if it makes sense to hide this behind a flag for now as we roll out this feature?

sraikund16 commented:
@eqy @syed-ahmed can I get a review on this?

eqy commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -983,6 +983,53 @@ class RingBuffer {
                    // will already be destroyed when we are in exit handlers
 };
 
+template <class T>
```
I'm not exactly sure what the purpose of this class is, is there explicit functionality that doesn't exist in an STL container? I see that vector is already being used here internally.

sraikund16 replied:
@eqy I can remove it. Originally thought it would be useful to have a class to handle max size and threading if needed but eventually moved over to making the stack thread-local instead

sraikund16 commented:
@eqy updated to use std::stack instead. May I have another look?

sraikund16 committed:
[Memento] Add PT2 to Memory Snapshot (#152707)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/152707

To add PT2 information to memory snapshot we piggyback off of the Kineto implementation using record_function similar to adding the user annotations. To do this we add the following:

1. Stack implementation that we instantiate to keep track of which compile context stack we are currently in (top element of the stack). The stack will be per device and thread-local since different threads of a process can be in different compile contexts at a given time. For this reason, we do not need to add mutexes to our stack impl since no two threads will touch a given stack

2. RecordFunction hooks to properly pipe the correct events to the compile context stack. These hooks are similar to the annotation ones in the fact that we just register them lazily and DO NOT unregister them. This is done out of convenience. In the future, we should save the handles and unregister them to minimize overhead after profiling is finished. As of now, we are registering this at the FUNCTION scope which is wide; however, we treat any function that does not start with "Torch-Compiled Region" as a no-op so we anticipate the difference in performance to be negligible during and after profiling.

3. Pipe compile context to pickle output

Test Plan: In D74039793, we add CompileContext to the visualizer and we see the following {F1977654658}

Reviewed By: basilwong

Differential Revision: D74028214

sraikund16 commented:
@eqy @syed-ahmed Can I get a stamp on this? Several teams are waiting on this feature. Its also guarded by a feature flag so it will be innocuous to most users.

sraikund16 commented:
@pytorchbot merge
